azure-iot-device/azure/iot/device/common/pipeline/config.py

```diff
@@ -77,3 +82,23 @@ def _sanitize_cipher(cipher):
             raise TypeError("Invalid type for 'cipher'")
 
         return cipher
+
+    @staticmethod
+    def _rectify_keep_alive(keep_alive):
+        if keep_alive and not isinstance(keep_alive, (int, float)):
+            raise TypeError("Invalid type for 'keep alive'. Permissible types are number.")
+
+        if keep_alive is not None and keep_alive <= 0:
+            # Not allowing a keep alive of 0 as this would mean frequent ping exchanges.
+            logger.error(
+                "'keep alive' can not be zero or negative. A default value of 'keep alive' will be used by the protocol."
+            )
+            keep_alive = None
+        elif keep_alive and keep_alive > constant.LOAD_BALANCER_LIMIT_SECS:
+            # This is Hub's maximum Load Balancer Limit before
+            logger.error(
+                "'keep alive' can not be more than 29 minutes. 'keep alive' will be set to max value of 29 minutes to continue."
+            )
```

Comment:
IMO, raise a ValueError here instead of just setting the value as the ceiling

Reply:
yes,

---
In reply to: [461187294](https://github.com/Azure/azure-iot-sdk-python/pull/635#discussion_r461187294) [](ancestors = 461187294)

Reply:
is it necessary for the user to know the bounds ?  I actually went back and forth with this as technically these are all correct values for keep alive but we have bounds simply because we have a load balancer... it is not a wrong value as such considering only the keep alive param but rather our load balancer will go wrong with these values...

---
In reply to: [461197621](https://github.com/Azure/azure-iot-sdk-python/pull/635#discussion_r461197621) [](ancestors = 461197621,461187294)